src/query/service/src/interpreters/interpreter_table_create.rs

```diff
@@ -144,7 +152,55 @@ impl CreateTableInterpreter {
 
     async fn create_table(&self) -> Result<PipelineBuildResult> {
         let catalog = self.ctx.get_catalog(self.plan.catalog.as_str())?;
-        catalog.create_table(self.build_request()?).await?;
+        catalog.create_table(self.build_request(None)?).await?;
+
+        if let Some(snapshot_loc) = self.plan.options.get(OPT_KEY_SNAPSHOT_LOCATION) {
+            let table = self
+                .ctx
+                .get_table(
+                    self.plan.catalog.as_str(),
+                    self.plan.database.as_str(),
+                    self.plan.table.as_str(),
+                )
+                .await?;
+            let fuse_table = FuseTable::try_from_table(table.as_ref())?;
```

Comment:
Putting the fuse table related here is a bit tricky?
Is it possible to check the table engine and get the stats from the table stats API:
```
fn table_statistics()
```

Reply:
Try to do this but if we modify will always read snapshot

Reply:
> Putting the fuse table related here is a bit tricky?
> 
> Is it possible to check the table engine and get the stats from the table stats API:
> 
> ```
> 
> fn table_statistics()
> 
> ```
> 
> 

And now if we use a not exists snapshot loc create table, it can succ. Will err only query it.

Reply:
Can we fix this with table API or does it in the internal table engine implementation not put the codes to the interpreter?

Reply:
Like this?

async fn table_statistics(&self) -> Result<Option<TableStatistics>> {
        let res = if let Some(snapshot) = self.read_table_snapshot().await? {
            Some(
                TableStatistics {
                    num_rows: Some(snapshot.summary.row_count),
                    data_size: Some(snapshot.summary.uncompressed_byte_size),
                    data_size_compressed: Some(snapshot.summary.compressed_byte_size),
                    index_size: Some(snapshot.summary.index_size),
                }
            )
        } else {
            let s = self.table_info.meta.statistics.clone();
            Some(TableStatistics {
                num_rows: Some(s.number_of_rows),
                data_size: Some(s.data_bytes),
                data_size_compressed: Some(s.compressed_data_bytes),
                index_size: Some(s.index_data_bytes),
            })
        };
        Ok(res)
    }

Reply:
I think we can do this. because create table XX (id int) snapshot_loc='yy' the 'yy' is from another table named YY and I can not get the table id of YY until the XX table create success.

https://github.com/datafuselabs/databend/issues/10406#issuecomment-1458071911

And the value is incorrect is because read stat in rule_fold_count_aggregate.rs.

How about when the precise_cardinality not equal 0, use this optimize?

```rust
if card != 0 {
    let mut scalars = agg.aggregate_functions;
    for item in scalars.iter_mut() {
        item.scalar = ScalarExpr::ConstantExpr(ConstantExpr {
            value: Literal::UInt64(card),
            data_type: Box::new(item.scalar.data_type()),
        });
    }
    let eval_scalar = EvalScalar { items: scalars };
    let dummy_table_scan = DummyTableScan;
    state.add_result(SExpr::create_unary(
        eval_scalar.into(),
        SExpr::create_leaf(dummy_table_scan.into()),
    ));
}
```

select min/max(id) from XX is ok, because of the column provider is read from the snapshot.

Reply:
After modify fold count:

```sql
mysql> select count(1) from tt_v2;
+----------+
| count(1) |
+----------+
|      100 |
+----------+
1 row in set (0.05 sec)
Read 100 rows, 0.00 B in 0.008 sec., 12.94 thousand rows/sec., 0.00 B/sec.

mysql> explain select count(1) from tt_v2;
+----------------------------------------------------------------------------------------------------------------------------------+
| explain                                                                                                                          |
+----------------------------------------------------------------------------------------------------------------------------------+
| EvalScalar                                                                                                                       |
| ├── expressions: [count(1) (#2)]                                                                                                 |
| ├── estimated rows: 1.00                                                                                                         |
| └── AggregateFinal                                                                                                               |
|     ├── group by: []                                                                                                             |
|     ├── aggregate functions: [count()]                                                                                           |
|     ├── estimated rows: 1.00                                                                                                     |
|     └── AggregatePartial                                                                                                         |
|         ├── group by: []                                                                                                         |
|         ├── aggregate functions: [count()]                                                                                       |
|         ├── estimated rows: 1.00                                                                                                 |
|         └── TableScan                                                                                                            |
|             ├── table: default.default.tt_v2                                                                                     |
|             ├── read rows: 100                                                                                                   |
|             ├── read bytes: 0                                                                                                    |
|             ├── partitions total: 1                                                                                              |
|             ├── partitions scanned: 1                                                                                            |
|             ├── pruning stats: [segments: <range pruning: 1 to 1>, blocks: <range pruning: 1 to 1, bloom pruning: 0 to 0>]       |
|             ├── push downs: [filters: [], limit: NONE]                                                                           |
|             ├── output columns: []                                                                                               |
|             └── estimated rows: 0.00                                                                                             |
+----------------------------------------------------------------------------------------------------------------------------------+
21 rows in set (0.04 sec)
Read 0 rows, 0.00 B in 0.003 sec., 0 rows/sec., 0.00 B/sec.

mysql> analyze table tt_v2;
Query OK, 0 rows affected (0.05 sec)

mysql> explain select count(1) from tt_v2;
+----------------------------------------+
| explain                                |
+----------------------------------------+
| EvalScalar                             |
| ├── expressions: [count(1) (#2)]       |
| ├── estimated rows: 1.00               |
| └── EvalScalar                         |
|     ├── expressions: [100]             |
|     ├── estimated rows: 1.00           |
|     └── DummyTableScan                 |
+----------------------------------------+
7 rows in set (0.04 sec)
Read 0 rows, 0.00 B in 0.004 sec., 0 rows/sec., 0.00 B/sec.

```

Reply:
Now we use operator from QueryContext to read_snapshot when create table.